OpenRA.Mods.Common/Scripting/Global/DateTimeGlobal.cs

```diff
@@ -20,11 +21,15 @@ namespace OpenRA.Mods.Common.Scripting
 	public class DateGlobal : ScriptGlobal
 	{
 		readonly TimeLimitManager tlm;
+		readonly int ticksPerSecond;
 
 		public DateGlobal(ScriptContext context)
 			: base(context)
 		{
 			tlm = context.World.WorldActor.TraitOrDefault<TimeLimitManager>();
+			var defaultGameSpeed = Game.ModData.Manifest.Get<GameSpeeds>().Speeds.FirstOrDefault(s => s.Key == "default");
+			var defaultTimestep = defaultGameSpeed.Value != null ? defaultGameSpeed.Value.Timestep : 40;
```

Comment:
Not sure if it is a good idea to hard-code `40` in here. Maybe use `new GameSpeed().Timestep`?

Reply:
Do we have a usecase for mods not defining a default gamespeed?
IMO we should just make this a requirement and not have the fallback at all.

Reply:
How about something like this (note: not tested ingame)?

This would avoid both hardcoding the "default" name and the fallback

```csharp
var gameSpeed = context.World.WorldActor.Info.TraitInfoOrDefault<MapOptionsInfo>()?.GameSpeed;
if (gameSpeed == null)
	throw new LuaException("Unable to determine the default game speed.");

if (!Game.ModData.Manifest.Get<GameSpeeds>().Speeds.TryGetValue(gameSpeed, out var info))
	throw new LuaException("Game speed '{0}' not defined.");

ticksPerSecond = 1000 / info.Timestep;
```

Reply:
I saw this too late for my update, but I'm not convinced it would be the right approach.

First of all, it would still kind of lead to https://github.com/OpenRA/OpenRA/pull/19297#discussion_r604356357

Furthermore, from what I've seen `MapOptions.GameSpeed` is only populated on actor creation, so fairly late.
At least in cases like shellmaps (possibly in other situations as well), that leads to `MapOptions.GameSpeed` not necessarily matching the actual gamespeed, so there's still a risk of a mismatch between what the modder expects vs. what the game actually does.

Reply:
Right, makes sense.

We've been going around in circles trying to avoid the mess that I made when I first implemented the feature (sorry...). How about this idea to cut through some of the mess?

* Change the speed definitions in mod.yaml to look something like:
   ```
   GameSpeeds:
   	DefaultSpeed: default
   	Speeds:
   		slowest:
   			Name: Slowest
   			Timestep: 80
   			OrderLatency: 2
   		# etc...
   ```
* Change the definition in `MapOptions` to
   ```csharp
   [Desc("Default game speed (leave empty to use the default defined in mod.yaml).")]
   public readonly string GameSpeed = null;
   ```
* Remove `Timestep` and `OrderLatency` completely from the `Global` class, and the hack updating these from LobbyCommands. Instead define a `string GameSpeed = null`.
* Change the `World` ctor to set `World.Timestep` and a new `World.OrderLatency` based on `gameSpeeds.Speeds[orderManager.LobbyInfo.GlobalSettings.GameSpeed ?? gameSpeeds.DefaultSpeed]`

This would hopefully centralize the game speed control so that the default specified in mod.yaml is always used unless the lobby specifies an override, also avoiding magic names.

Reply:
OK, combined with your other comment it makes sense.
It did work locally on the RA shellmap, so I'll update accordingly for now.

Reply:
Ugh, crossed comments again!

What do you think about my latest suggestion?

Reply:
That looks a lot better than what we have. I'll see if I run into any issues implementing that and go to IRC if I do to avoid more comment-crossing.

Reply:
Another quick thought to improve things further: we might be able to introduce a new interface (`IGameSpeed`?) that the world ctor can use to query `MapOptions` (which is in Mods.Common, so not available to directly query) for its game speed. This would allow us to avoid `GlobalSettings.GameSpeed` completely and the serverside hack to set it

We'd then fall back to `gameSpeeds.DefaultSpeed` if there isn't exactly one trait implementing `IGameSpeed` or it returns a null value.

Note that i'm going to be afk/mobile for the next hour or two, so may not be able to respond on IRC immediately.

Reply:
A simpler solution that should give most of the same benefits will be to ignore/remove `GlobalSettings.GameSpeed` and the `LobbyCommands` hack as above, but instead of a new interface we just hardcode the "gamespeed" option in the world ctor with something like:
```csharp
var gameSpeedName = self.World.LobbyInfo.GlobalSettings.OptionOrDefault("gamespeed", gameSpeeds.DefaultSpeed);
var gameSpeed = gameSpeeds.Speeds[gameSpeedName];
```